File: AppRemover/src/main/java/com/vs_unusedappremover/AppEntry.java
package com.vs_unusedappremover;

import android.content.pm.ApplicationInfo;

public class AppEntry {

    public static enum RanIn {
        UNKNOWN(0),
        BACKGROUND(1),
        FOREGROUND(2);

        public final int id;

        public static RanIn byId(int id) {
            for (RanIn v : values()) {
                if (v.id == id) return v;
            }
            throw new IllegalArgumentException("No item with such id");
        }

        private RanIn(int id) {
            this.id = id;
        }
    }

    public String label;
    public long size;
    public ApplicationInfo info;
    public long lastUsedTime;
    public RanIn ranIn;
    public long installTime;
    public float rating;
    public String downloadCount;
    public boolean notifyAbout = true;
}

File: AppRemover/src/main/java/com/vs_unusedappremover/AppEntry.kt
package com.vs_unusedappremover

import android.content.pm.ApplicationInfo
import android.net.Uri


data class AppEntry (var info: ApplicationInfo,
                var label: String,
                var size: Long = PackageSize.UNKNOWN,
                var installTime: Long = 0,
                var lastUsedTime: Long = 0,
                var ranIn: RanIn = RanIn.UNKNOWN,
                var rating: Float = 0F,
                var notifyAbout: Boolean = true,
                var downloadCount: String? = null) {

    enum class RanIn constructor(val id: Int) {
        UNKNOWN(0),
        BACKGROUND(1),
        FOREGROUND(2);


        companion object {

            fun byId(id: Int): RanIn {
                values()
                        .filter { it.id == id }
                        .forEach { return it }
                throw IllegalArgumentException("No item with such id")
            }
        }
    }

    fun buildUrl(): Uri {
        val packageName: String = info.packageName
        return Uri.parse(AppIcon.SCHEME + "://" + packageName)
    }

    companion object {
        val byLabel = compareBy(AppEntry::label)
        val bySize = compareByDescending(AppEntry::size)
        val byTimeUnused = Comparator<AppEntry> { e1, e2 ->
            val time1 = Math.max(e1.installTime, e1.lastUsedTime)
            val time2 = Math.max(e2.installTime, e2.lastUsedTime)
            if (time1 == time2) 0 else if (time1 < time2) -1 else 1
        }.then(byLabel)
    }
}
